### ISSUE TITLE
Parse the HTML to obtain the simulation third party licenses

### ISSUE THREAD

**Issue by samreid** [phetsims/chipper#220]
See https://github.com/phetsims/chipper/issues/207

Instead of a separate process and output file for the sim's 3rd party licenses, we are going to use the same plain js objects as reported in the HTML file.

**Comment by samreid** [comment]
These are being parsed wonderfully in grunt report-third-party. Reassigned to @pixelzoom for review.

**Comment by pixelzoom** [comment]
I built gravity-and-orbits and wanted to generate a 3rd-party report for its HTML file only.  So I tried this:

```js
% cd gravity-and-orbits
% grunt
% grunt report-third-party ./build
Warning: Task "./build" not found. Use --force to continue.

Aborted due to warnings.
%
```

Then I tried this:

```js
% mkdir /tmp/html-files
% cp build/gravity-and-orbits_en.html /tmp/html-files
% grunt report-third-party
Warning: Task "/tmp/html-files/" not found. Use --force to continue.

Aborted due to warnings.
%
```

What am I doing wrong?

**Comment by pixelzoom** [comment]
Ah... No output dir perhaps?  Let's try:

```
% grunt report-third-party ~/Downloads/html-files/ .
Warning: Task "/Users/cmalley/Downloads/html-files/" not found. Use --force to continue.

Aborted due to warnings.
%
```

**Comment by pixelzoom** [comment]
Ah, so it's reporting `Task "/Users/cmalley/Downloads/html-files/" not found`, so thinks the args are task names. What are the options for specifying the required arguments that are noted in `grunt --help`.

**Comment by samreid** [comment]
Required arguments are passed using a colon:
http://gruntjs.com/api/grunt.task

**Comment by pixelzoom** [comment]
`grunt --help` says:

```js
 report-third-party  Given a directory of HTML files (required argument),  
                         report the third party code, images, audio, etc. used 
                         in the group and commit the changes to                
                         sherpa/third-party-licenses.json. If the              
                         "--copy-from-build" option is given, it will iterate  
                         over the active-runnables and copy all the files to   
                         the report HTML directory (required argument) 
```

This indicates 2 required args, the input directory for the .html files, and the output directory for the report (assuming that's what's mean by "the report HTML directory").  I was unable to get the report to appear anywhere other than in my working copy of sherpa. 

For example:

```js
% ls  ls /Users/cmalley/Downloads/html-files/
gravity-and-orbits_en.html
% grunt report-third-party:/Users/cmalley/Downloads/html-files/:/Users/cmalley/Downloads
% ls /Users/cmalley/Downloads/third-party-licenses.md
ls: /Users/cmalley/Downloads/third-party-licenses.md: No such file or directory
%
```

If "the report HTML directory" is only required when the `--copy-from-build` option is present, that's not clear , and it doesn't make a lot of sense - why would I be able to specify an output dir only when using copying all .html files from my working copy, but not be able to do that when I have a specific set of .html files.  Seems like the opposite should be the case.

**Comment by samreid** [comment]
Let me have a try at fixing that documentation.

**Comment by samreid** [comment]
In writing the documentation, I realized that Windows developers could have trouble passing options to grunt on the command line if the path contains a colon (like C:/mydirectory).  We may need to change this to a "required" "optional" argument to work around this issue.

**Comment by pixelzoom** [comment]
On bash, I also ran into trouble when trying to use '~' in paths, e.g. ~/Downloads/html-files.  Didn't try '..' or '.' to see if they were resolved properly.

**Comment by samreid** [comment]
Proposed documentation:

```
report-third-party

This task is used to create a report of third-party resources (code, images, audio, etc) 
used in a subset of PhET simulations by reading the license information in the built HTML file.
The report is copied over sherpa/third-party-licenses.json.
--path (required argument) the path to the directory containing HTML files which will 
be reported on.  Note: this is a required argument but we chose not to use the grunt
required argument syntax (a colon) since it would cause problems for Windows 
developers who have colons in the pathnames.
--copy-from-build (optional flag) If this flag is supplied, the task iterate over the active-runnables 
and copy each built HTML file into the directory specified with --path before running the
report.  If any HTML files are missing, the report will fail.  Active runnables without titles
(such as mikes-playground) will be skipped.  Before using this flag, the developer 
should run `grunt-all.sh` to make sure all of the HTML files are up-to-date.
```

**Comment by samreid** [comment]
If the proposed documentation sounds good to @pixelzoom I'll implement it as described.

**Comment by pixelzoom** [comment]
Questions/comments:

(1) No option for where the report is written? What if I have chipper checked out so I can run the report on a set of html files, but don't have sherpa checked out?

(2) Instead of `--copy-from-build`, how about `--active-runnables`?

(3) `--path` seems quite generic. If we're going to go generic (which might be ok), then maybe `--inputPath`, so we can also have `--outputPath`?

(4) Doc says "Active runnables without titles (such as mikes-playground) will be skipped."  Does "without titles" mean without simTitleStringKey in package.json? (If so, mikes-playground does have a title.)  Why are they skipped - not possible to use repo name instead? Are they skipped silently or with a warning?

**Comment by pixelzoom** [comment]
(5) If I use `--copy-from-build`, does it clean the directory before copying files? Or might I end up with some html files that were hanging around from the last time I used the same directory?

**Comment by samreid** [comment]
Revised documentation, should address (1) (2) (3) (4) and (5) above.

```
report-third-party

This task is used to create a report of third-party resources (code, images, audio, etc) 
used in a subset of PhET simulations by reading the license information in the built HTML file.
The report is copied over sherpa/third-party-licenses.json.
--inputPath (required argument) the path to the directory containing HTML files which will 
be reported on.  Note: this is a required argument but we chose not to use the grunt
required argument syntax (a colon) since it would cause problems for Windows 
developers who have colons in the pathnames.
--outputPath (required argument) the path to a file where the report should be written.  This will silently overwrite an existing file, if there is one.
--active-runnables (optional flag, boolean) If this flag is supplied, the task iterates over the active-runnables 
and copies each built HTML file into the directory specified with --inputPath before
running the report.  If any HTML files are missing, the report will fail.  Before using 
this flag, the developer should run `grunt-all.sh` to make sure all of the HTML files 
are up-to-date.  The directory is neither automatically created nor automatically 
cleaned, this is the responsibility of the developer.  (Note that if you fail to manually clean the directory, you may end up with stale HTML files).
```

In response to (4), I am seeing these simulations without a `<title>` attribute in the html 

> title not found for /Users/samreid/temp/blast_en.html
title not found for /Users/samreid/temp/litmus_en.html
title not found for /Users/samreid/temp/mikes-playground_en.html
title not found for /Users/samreid/temp/ph-scale-basics_en.html
title not found for /Users/samreid/temp/states-of-matter-basics_en.html

We should probably find out why there are no titles and include reports for them.  I'll create a separate issue for that shortly.  Reassigned to @pixelzoom to review the proposed documentation.  If it is good, reassign to me and I'll implement with that spec.

**Comment by pixelzoom** [comment]
The description says:
> The report is copied over sherpa/third-party-licenses.json.

then it says:
> --outputPath (required argument) the path to a file where the report should be written.  This will silently overwrite an existing file, if there is one.

Both of these together seems dubious, it says that the report will be written to 2 places.  Is `--outputPath` in fact optional, and sherpa/third-party-licenses.json the default location for the report?

**Comment by pixelzoom** [comment]
> Note: this is a required argument but we chose not to use the grunt required argument syntax (a colon) since it would cause problems for Windows developers who have colons in the pathnames.

This is documented only for inputPath, and the same is true for outputPath. And I don't think this is necessary/desirable in the description.  I recommend documenting this at the `grunt.option` calls that read the inputPath and outputPath option.

**Comment by samreid** [comment]
> Both of these together seems dubious, it says that the report will be written to 2 places.

I left a dangling reference, will remove it in the next version.  The --outputPath will be a required argument.  Also, it just crossed my mind to rename inputPath=>input and outputPath=>output.

I'll move out the documentation for required argument as you suggested.  We'll have to remember to add that to the docs when it's implementation time.

```
report-third-party

This task is used to create a report of third-party resources (code, images, audio, etc) 
used in a subset of PhET simulations by reading the license information in built HTML files.
--input (required argument) the path to the directory containing HTML files which will 
be reported on.
--output (required argument) the path to a file where the report should be written.  The 
file is in markdown syntax and the *.md suffix is recommended.  This will silently 
overwrite an existing file, if there is one.
--active-runnables (optional flag, boolean) If this flag is supplied, the task iterates 
over the active-runnables and copies each built HTML file into the directory 
specified with --input before running the report.  If any HTML files are missing, 
the report will fail.  Before using  this flag, the developer should run `grunt-all.sh` 
to make sure all of the HTML files are up-to-date.  The directory is neither automatically
created nor automatically cleaned, this is the responsibility of the developer.  
(Note that if you fail to manually clean the directory, you may end up with stale HTML files).
```

**Comment by pixelzoom** [comment]
Proposed help description looks good. Reminder to document why we're not using grunt arg syntax at the calls to `grunt.option( 'input' )` and `grunt.option( 'output' )`.

**Comment by samreid** [comment]
I added the docs as above, and documented the grunt.option('input') and output are using the optional arg syntax.  I also assert that both paths are provided and upconvert the optional option `active-runnables` to a boolean.  I revised the task to use these args properly and tested with a one-sim report and an `active-runnables` report.  Both looked good and I published the active runnables report to sherpa at https://github.com/phetsims/sherpa/blob/master/third-party-licenses.md

I'll skim my change set to see if anything is amiss, lint again and commit shortly.
